File: obdii/core/OBDConnectionManager.swift
import Combine
import CoreBluetooth
/**
 * __Final Project__
 * Jim Mittler
 * 19 November 2025
 *
 * Main OBD-II connection manager singleton
 *
 * Manages vehicle communication via SwiftOBD2 library, connection lifecycle,
 * and continuous PID data streaming. Integrates with PIDInterestRegistry for
 * demand-driven polling. Publishes connection state, diagnostic codes, fuel
 * status, MIL status, and per-PID statistics. Supports Bluetooth and Wi-Fi adapters.
 */
import Foundation
import SwiftOBD2

@MainActor
final class OBDConnectionManager: ObservableObject {

  enum ConnectionState: Equatable {
    case disconnected
    case connecting
    case connected
    case failed(Error)

    static func == (lhs: ConnectionState, rhs: ConnectionState) -> Bool {
      switch (lhs, rhs) {
      case (.disconnected, .disconnected),
        (.connecting, .connecting),
        (.connected, .connected):
        return true
      case (.failed(let le), .failed(let re)):
        return le.localizedDescription == re.localizedDescription
      default:
        return false
      }
    }

    var isFailed: Bool {
      if case .failed = self { return true }
      return false
    }
  }
  // Whether to query the adapter/vehicle for supported PIDs and filter Mode 01 accordingly.
  private let querySupportedPids = true

  static let shared = OBDConnectionManager()

  // connection status
  @Published var connectionState: ConnectionState = .disconnected
  // Current DTCs (nil = not yet received, [] = loaded but none).
  @Published var troubleCodes: [TroubleCodeMetadata]? = nil
  // FI/O2 fuel system status (nil = not yet received).
  @Published var fuelStatus: [StatusCodeMetadata?]? = nil
  // MIL status (nil = not yet received).
  @Published var MILStatus: Status? = nil
  // Bluetooth peripheral name, or nil for Wi-Fi/Demo/none.
  @Published var connectedPeripheralName: String? = nil
  // Per-PID statistics for live gauge values.
  @Published private(set) var pidStats: [OBDCommand: PIDStats] = [:]

  // our main structure for storing PID stats
  struct PIDStats: Equatable {
    var pid: OBDCommand
    var latest: MeasurementResult
    var min: Double
    var max: Double
    var sampleCount: Int

    init(pid: OBDCommand, measurement: MeasurementResult) {
      self.pid = pid
      self.latest = measurement
      self.min = measurement.value
      self.max = measurement.value
      self.sampleCount = 1
    }

    mutating func update(with measurement: MeasurementResult) {
      let value = measurement.value
      latest = measurement
      if value < min { min = value }
      if value > max { max = value }
      sampleCount &+= 1
    }
  }

  // our gateway to the SwiftOBD library to do the comms
  private var obdService: OBDService
  // General long-lived subscriptions (registry, config, etc.)
  private var managerCancellables = Set<AnyCancellable>()
  // Subscriptions tied to the current OBDService instance.
  private var serviceCancellables = Set<AnyCancellable>()
  // Subscriptions for the continuous updates stream.
  private var streamCancellables = Set<AnyCancellable>()
  // Mode 1 and other supported PIDs reported by the vehicle.
  private var supportedPids: [OBDCommand] = []
  // Last set of PIDs that were actually being streamed.
  private var lastStreamingPIDs: Set<OBDCommand> = []

  private init() {
    obdService = OBDService(
      connectionType: ConfigData.shared.connectionType,
      host: ConfigData.shared.wifiHost,
      port: UInt16(ConfigData.shared.wifiPort)
    )

    bindServiceMirrors()
    bindInterestRegistry()
    bindUnitChanges()
  }
  // Bind OBDService publishers (per-instance).
  private func bindServiceMirrors() {
    serviceCancellables.removeAll()

    // Mirror the connected peripheral name
    obdService.$connectedPeripheral
      .map { $0?.name }
      .removeDuplicates()
      .receive(on: DispatchQueue.main)
      .sink { [weak self] name in
        self?.connectedPeripheralName = name
      }
      .store(in: &serviceCancellables)

    // Mirror the OBDService connection state into our own state machine
    obdService.$connectionState
      .removeDuplicates()
      .receive(on: DispatchQueue.main)
      .sink { [weak self] serviceState in
        self?.handleServiceConnectionState(serviceState)
      }
      .store(in: &serviceCancellables)
  }
  // Listen for demand-driven interest changes from PIDInterestRegistry.
  private func bindInterestRegistry() {
    PIDInterestRegistry.shared.$interested
      .removeDuplicates()
      .sink { [weak self] interestedSet in
        guard let self else { return }

        // If connected, restart continuous updates with the new interest set
        if connectionState == .connected {
          restartContinuousUpdates(with: interestedSet)
        } else {
          // Not connected: clear stream state so we start clean once connected
          streamCancellables.removeAll()
          lastStreamingPIDs = []
        }
      }
      .store(in: &managerCancellables)
  }
  // Listen for unit changes and reset stats + restart streams as appropriate.
  private func bindUnitChanges() {
    ConfigData.shared.$units
      .removeDuplicates()
      .receive(on: DispatchQueue.main)
      .sink { [weak self] _ in
        guard let self else { return }
        let interestedSet = PIDInterestRegistry.shared.interested

        if connectionState == .connected {
          obdInfo(
            "Units changed to \(ConfigData.shared.units.rawValue); restarting continuous updates and resetting stats.",
            category: .service)
          resetAllStats()
          lastStreamingPIDs = []
          restartContinuousUpdates(with: interestedSet)
        } else {
          resetAllStats()
        }
      }
      .store(in: &managerCancellables)
  }

  private func clearForTerminalState() {
    streamCancellables.removeAll()
    lastStreamingPIDs = []
    pidStats.removeAll()
    fuelStatus = nil
    MILStatus = nil
    troubleCodes = nil
    connectedPeripheralName = nil
  }

  private func handleServiceConnectionState(_ serviceState: SwiftOBD2.ConnectionState) {
    switch serviceState {
    case .disconnected:
      clearForTerminalState()
      connectionState = .disconnected
      obdDebug("Mirrored service disconnect to manager state and cleared data.", category: .service)

    case .error:
      clearForTerminalState()
      let err = OBDServiceError.notConnectedToVehicle
      connectionState = .failed(err)
      obdError(
        "Service reported error \(err); mapped to .failed and cleared data.", category: .service)

    case .connecting:
      if connectionState != .connecting {
        connectionState = .connecting
      }

    case .connectedToAdapter, .connectedToVehicle:
      if connectionState != .connected {
        connectionState = .connected
      }
    }
  }
  // Called when host/port/connectionType changes.
  // Disconnects if needed, recreates the OBDService with new settings, and re-binds its publishers.
  func updateConnectionDetails() {
    if connectionState != .disconnected {
      disconnect()
    }

    obdService = OBDService(
      connectionType: ConfigData.shared.connectionType,
      host: ConfigData.shared.wifiHost,
      port: UInt16(ConfigData.shared.wifiPort)
    )

    bindServiceMirrors()
    obdInfo("OBD Service re-initialized with new settings.", category: .service)
  }

  func connect() async {
    guard connectionState == .disconnected || connectionState.isFailed else {
      obdWarning("Connection attempt ignored, already connected or connecting.", category: .service)
      return
    }

    connectionState = .connecting

    do {
      _ = try await obdService.startConnection(
        preferedProtocol: .protocol6,
        timeout: 30,
        querySupportedPIDs: querySupportedPids
      )

      supportedPids = await obdService.getSupportedPIDs()

      connectionState = .connected
      connectedPeripheralName = obdService.connectedPeripheral?.name
      obdInfo("OBD-II connected successfully.", category: .service)

      // Start continuous updates using the current interest set (may be empty)
      startContinuousOBDUpdates(with: PIDInterestRegistry.shared.interested)

    } catch {
      let message = error.localizedDescription
      connectionState = .failed(error)
      connectedPeripheralName = nil
      obdError("OBD-II connection failed: \(message)", category: .service)
    }
  }

  func disconnect() {
    obdService.stopConnection()
    // Do not clear managerCancellables here (those are long-lived)
    streamCancellables.removeAll()
    clearForTerminalState()
    connectionState = .disconnected
    obdInfo("OBD-II disconnected.", category: .service)
  }

  private func resetAllStats() {
    pidStats = pidStats.mapValues { existing in
      var reset = existing
      reset.min = existing.latest.value
      reset.max = existing.latest.value
      reset.sampleCount = 1
      return reset
    }
    obdInfo("All PID stats reset (min/max/sampleCount).", category: .service)
  }

  func stats(for pid: OBDCommand) -> PIDStats? {
    pidStats[pid]
  }

  private func startContinuousOBDUpdates(with interestedPIDs: Set<OBDCommand>) {
    startContinuousOBDUpdatesInternal(with: interestedPIDs)
  }

  private func startContinuousOBDUpdatesInternal(with interestedPIDs: Set<OBDCommand>) {
    var enabledNow = interestedPIDs

    // Filter by supported PIDs if requested
    if querySupportedPids {
      let supportedMode1: Set<OBDCommand> = Set(
        supportedPids.compactMap { cmd in
          if case .mode1(let m) = cmd { return .mode1(m) }
          return nil
        }
      )

      let filtered = Set(
        interestedPIDs.filter { cmd in
          switch cmd {
          case .mode1:
            return supportedMode1.contains(cmd)
          default:
            return true
          }
        })

      let removed = interestedPIDs.subtracting(filtered)
      if !removed.isEmpty {
        obdDebug("Removing unsupported mode1 PIDs: \(removed)", category: .service)
      }

      enabledNow = filtered
    }

    // No commands to monitor
    guard !enabledNow.isEmpty else {
      obdInfo("No interested PIDs to monitor.", category: .service)
      if !lastStreamingPIDs.isEmpty {
        streamCancellables.removeAll()
        lastStreamingPIDs = []
      }
      return
    }

    // If nothing changed, do nothing
    if enabledNow == lastStreamingPIDs {
      obdInfo("Interested PIDs unchanged; not restarting continuous updates.", category: .service)
      return
    }

    // Replace only the stream subscription
    streamCancellables.removeAll()

    let commands = Array(enabledNow)
    lastStreamingPIDs = enabledNow

    obdInfo(
      "Starting continuous updates for \(commands.count) PIDs (demand-driven).", category: .service)

    obdService
      .startContinuousUpdates(commands, unit: ConfigData.shared.units, interval: 1)
      .receive(on: DispatchQueue.main)
      .sink(
        receiveCompletion: { completion in
          if case .failure(let error) = completion {
            obdError(
              "Continuous OBD updates failed: \(error.localizedDescription)", category: .service)
          }
        },
        receiveValue: { [weak self] measurements in
          self?.handleUpdateBatch(measurements)
        }
      )
      .store(in: &streamCancellables)
  }

  // deal with a batch of results
  private func handleUpdateBatch(_ batch: [OBDCommand: DecodeResult]) {
    for (command, decode) in batch {
      switch command {

      case .mode1(let pid):
        switch pid {
        // fuel status and MIL status are special cases that we store separately
        case .fuelStatus:
          if let codes = decode.codeResult {
            fuelStatus = codes
          }
        case .status:
          if let status = decode.statusResult {
            MILStatus = status
          }
        // gauges
        default:
          if let measurement = decode.measurementResult {
            let key: OBDCommand = .mode1(pid)
            var stats = pidStats[key] ?? PIDStats(pid: key, measurement: measurement)
            stats.update(with: measurement)
            pidStats[key] = stats
          }
        }

      // GM specific pids
      case .GMmode22:
        if let measurement = decode.measurementResult {
          let key: OBDCommand = command
          var stats = pidStats[key] ?? PIDStats(pid: key, measurement: measurement)
          stats.update(with: measurement)
          pidStats[key] = stats
        }

      // DTC pid
      case .mode3(let m3):
        switch m3 {
        case .GET_DTC:
          if let dtcs = decode.troubleCodesByECU,
            let engine = dtcs[SwiftOBD2.ECUID.engine]
          {
            // Real payload, possibly empty
            troubleCodes = engine
          } else {
            // Loaded, but no codes
            troubleCodes = []
          }
        }

      default:
        continue
      }
    }
  }

  private func restartContinuousUpdates(with interestedPIDs: Set<OBDCommand>) {
    streamCancellables.removeAll()
    lastStreamingPIDs = []
    startContinuousOBDUpdatesInternal(with: interestedPIDs)
  }
}

File: flutter_obdii/lib/core/obd_connection_manager.dart
// Port of OBDConnectionManager.swift — Jim Mittler
// Main OBD-II connection manager singleton.
// Manages vehicle communication via flutter_obd2 library, connection lifecycle,
// and continuous PID data streaming. Integrates with PIDInterestRegistry for
// demand-driven polling. Publishes connection state, diagnostic codes, fuel
// status, MIL status, and per-PID statistics.

import 'dart:async';
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:flutter_blue_plus/flutter_blue_plus.dart';
import 'package:flutter_obd2/flutter_obd2.dart' as obd2lib;
import 'package:permission_handler/permission_handler.dart';

import 'config_data.dart';
import 'logger.dart';
import 'pid_interest_registry.dart';

// ─────────────────────────────────────────────
// MARK: ConnectionState
// ─────────────────────────────────────────────

enum OBDConnectionState {
  disconnected,
  connecting,
  connectedToAdapter,
  settingUpVehicle,
  connected,
  failed,
}

// ─────────────────────────────────────────────
// MARK: PIDStats
// ─────────────────────────────────────────────

class PIDStats {
  final String pid;
  final obd2lib.MeasurementResult latest;
  final double min;
  final double max;
  final int sampleCount;

  PIDStats({required this.pid, required this.latest})
      : min = latest.value,
        max = latest.value,
        sampleCount = 1;

  // Private full constructor used by copyWith
  const PIDStats._({
    required this.pid,
    required this.latest,
    required this.min,
    required this.max,
    required this.sampleCount,
  });

  /// Returns a NEW PIDStats instance with accumulated min/max/sampleCount.
  /// Using a new object (not mutation) is critical: GaugesViewModel's
  /// _tilesEqual() compares stats by value, but when the same object is
  /// mutated both old and new tiles see identical fields and updates are
  /// silently dropped.
  PIDStats copyWith(obd2lib.MeasurementResult measurement) {
    final v = measurement.value;
    return PIDStats._(
      pid: pid,
      latest: measurement,
      min: v < min ? v : min,
      max: v > max ? v : max,
      sampleCount: sampleCount + 1,
    );
  }

  @override
  bool operator ==(Object other) =>
      other is PIDStats &&
      other.pid == pid &&
      other.latest.value == latest.value &&
      other.min == min &&
      other.max == max &&
      other.sampleCount == sampleCount;

  @override
  int get hashCode => Object.hash(pid, latest.value, min, max, sampleCount);
}


// ─────────────────────────────────────────────
// MARK: Provider protocols (replaces Swift protocols)
// ─────────────────────────────────────────────

abstract class PidStatsProviding {
  Map<String, PIDStats> get pidStats;
  PIDStats? statsFor(String pidCommand);
  Stream<Map<String, PIDStats>> get pidStatsStream;
}

abstract class DiagnosticsProviding {
  List<obd2lib.TroubleCodeMetadata>? get troubleCodes;
  Stream<List<obd2lib.TroubleCodeMetadata>?> get diagnosticsStream;
}

abstract class FuelStatusProviding {
  List<obd2lib.StatusCodeMetadata?>? get fuelStatus;
  Stream<List<obd2lib.StatusCodeMetadata?>?> get fuelStatusStream;
}

abstract class MilStatusProviding {
  obd2lib.Status? get milStatus;
  Stream<obd2lib.Status?> get milStatusStream;
}

abstract class OBDConnectionControlling {
  OBDConnectionState get connectionState;
  void updateConnectionDetails();
  Future<void> connect();
  void disconnect();
  Stream<OBDConnectionState> get connectionStateStream;
}

// ─────────────────────────────────────────────
// MARK: OBDConnectionManager singleton
// ─────────────────────────────────────────────

class OBDConnectionManager extends ChangeNotifier
    implements
        PidStatsProviding,
        DiagnosticsProviding,
        FuelStatusProviding,
        MilStatusProviding,
        OBDConnectionControlling {
  static final OBDConnectionManager instance = OBDConnectionManager._();
  OBDConnectionManager._();
  static const bool _querySupportedPids = true;

  /// When non-null (tests only), skips the real FlutterBluePlus lookup.
  @visibleForTesting
  static Future<bool> Function()? debugBluetoothIsSupportedOverride;

  /// When non-null (tests only), skips waiting on [FlutterBluePlus.adapterState].
  @visibleForTesting
  static Future<BluetoothAdapterState> Function()?
      debugBluetoothInitialAdapterStateOverride;

  // ── Published state ─────────────────────────────

  @override
  OBDConnectionState connectionState = OBDConnectionState.disconnected;

  /// Current DTCs. null = not yet received, [] = loaded but none.
  @override
  List<obd2lib.TroubleCodeMetadata>? troubleCodes;

  /// FI/O2 fuel system status. null = not yet received.
  @override
  List<obd2lib.StatusCodeMetadata?>? fuelStatus;

  /// MIL/Check Engine status. null = not yet received.
  @override
  obd2lib.Status? milStatus;

  /// Bluetooth peripheral name, or null for WiFi/Demo.
  String? connectedPeripheralName;

  @override
  Map<String, PIDStats> pidStats = {};

  // ── Internal ─────────────────────────────────────

  obd2lib.Obd2Service? _obdService;
  StreamSubscription? _connectionStateSub;
  StreamSubscription? _dataSub;
  StreamSubscription? _interestSub;
  StreamSubscription? _unitsSub;
  Set<String> _lastStreamingPids = {};
  Set<String> _supportedMode1Pids = {};
  bool _hasSupportedMode1Snapshot = false;
  int _connectionAttemptId = 0;

  // ── Notifiers (replaces Combine @Published) ──────

  final _connStateNotifier =
      ValueNotifier<OBDConnectionState>(OBDConnectionState.disconnected);
  final _pidStatsNotifier = ValueNotifier<Map<String, PIDStats>>({});

  @override
  Stream<OBDConnectionState> get connectionStateStream =>
      _valueNotifierStream(_connStateNotifier);

  @override
  Stream<Map<String, PIDStats>> get pidStatsStream =>
      _valueNotifierStream(_pidStatsNotifier);

  @override
  Stream<List<obd2lib.TroubleCodeMetadata>?> get diagnosticsStream =>
      _changeNotifierStream(() => troubleCodes);

  @override
  Stream<List<obd2lib.StatusCodeMetadata?>?> get fuelStatusStream =>
      _changeNotifierStream(() => fuelStatus);

  @override
  Stream<obd2lib.Status?> get milStatusStream =>
      _changeNotifierStream(() => milStatus);

  // ── Bootstrap ────────────────────────────────────

  void initialize() {
    _rebuildService();
    _bindInterestRegistry();
    _bindUnitChanges();
  }

  void _rebuildService() {
    final config = ConfigData.instance;
    final libConn = _toLibConnType(config.connectionType);
    _obdService = obd2lib.Obd2Service(
      connectionType: libConn,
      host: config.wifiHost,
      port: config.wifiPort,
    );
    _connectionStateSub?.cancel();
    _connectionStateSub =
        _obdService!.connectionStatePublisher.listen(_handleServiceState);
  }

  void _bindInterestRegistry() {
    _interestSub?.cancel();
    _interestSub =
        PidInterestRegistry.instance.interestedStream.listen((interested) {
      if (connectionState == OBDConnectionState.connected) {
        _restartContinuousUpdates(interested);
      } else {
        _dataSub?.cancel();
        _lastStreamingPids = {};
      }
    });
  }

  void _bindUnitChanges() {
    _unitsSub?.cancel();
    _unitsSub = ConfigData.instance.unitsStream.listen((_) {
      if (connectionState == OBDConnectionState.connected) {
        obdInfo(
          'Units changed to ${ConfigData.instance.units.name}; resetting stats and restarting updates.',
          category: LogCategory.service,
        );
        _resetAllStats();
        _lastStreamingPids = {};
        _restartContinuousUpdates(PidInterestRegistry.instance.interested);
      } else {
        _resetAllStats();
      }
    });
  }

  // ── Connection ───────────────────────────────────

  @override
  Future<void> connect() async {
    if (_shouldIgnoreConnectBecauseAlreadyConnecting()) return;

    final attemptId = ++_connectionAttemptId;
    if (_obdService == null) {
      _rebuildService();
    }
    final service = _obdService!;
    _setConnectionState(OBDConnectionState.connecting);

    final config = ConfigData.instance;
    if (config.connectionType == ConnectionType.bluetooth) {
      final ready = await _prepareBluetoothForConnection(attemptId, service);
      if (!ready) return;
    }

    await _establishObdConnectionSession(attemptId, service);
  }

  bool _shouldIgnoreConnectBecauseAlreadyConnecting() {
    if (connectionState == OBDConnectionState.connected ||
        connectionState == OBDConnectionState.connecting) {
      obdWarning('Connection attempt ignored, already connected or connecting.',
          category: LogCategory.service);
      return true;
    }
    return false;
  }

  /// Sets [failed] if this connection attempt is still current; otherwise no-op.
  void _markBluetoothPrepFailedIfStillActive(
    int attemptId,
    obd2lib.Obd2Service service,
  ) {
    if (!_isActiveConnectionAttempt(attemptId, service)) return;
    _setConnectionState(OBDConnectionState.failed);
  }

  Future<bool> _readFlutterBluePlusIsSupported() async {
    final override = debugBluetoothIsSupportedOverride;
    if (override != null) return override();
    return FlutterBluePlus.isSupported;
  }

  Future<BluetoothAdapterState> _readFlutterBluePlusInitialAdapterState() async {
    final override = debugBluetoothInitialAdapterStateOverride;
    if (override != null) return override();
    return FlutterBluePlus.adapterState.first;
  }

  Future<bool> _bluetoothPermissionsAndHardwareReady(
    int attemptId,
    obd2lib.Obd2Service service,
  ) async {
    if (!await _ensureBluetoothPermissions()) {
      _markBluetoothPrepFailedIfStillActive(attemptId, service);
      return false;
    }
    if (!_isActiveConnectionAttempt(attemptId, service)) return false;
    if (!await _readFlutterBluePlusIsSupported()) {
      _markBluetoothPrepFailedIfStillActive(attemptId, service);
      return false;
    }
    return _isActiveConnectionAttempt(attemptId, service);
  }

  Future<bool> _ensureBluetoothAdapterOnForSession(
    int attemptId,
    obd2lib.Obd2Service service,
  ) async {
    final adapterState = await _readFlutterBluePlusInitialAdapterState();
    if (!_isActiveConnectionAttempt(attemptId, service)) return false;
    if (adapterState == BluetoothAdapterState.on) return true;

    try {
      await FlutterBluePlus.turnOn();
    } catch (_) {
      _markBluetoothPrepFailedIfStillActive(attemptId, service);
      return false;
    }
    return _isActiveConnectionAttempt(attemptId, service);
  }

  /// Returns false if [connect] should stop (failure or superseded attempt).
  Future<bool> _prepareBluetoothForConnection(
    int attemptId,
    obd2lib.Obd2Service service,
  ) async {
    if (!await _bluetoothPermissionsAndHardwareReady(attemptId, service)) {
      return false;
    }
    return _ensureBluetoothAdapterOnForSession(attemptId, service);
  }

  Future<void> _establishObdConnectionSession(
    int attemptId,
    obd2lib.Obd2Service service,
  ) async {
    try {
      await _runStartConnectionHandshake(attemptId, service);
    } catch (e) {
      _handleStartConnectionFailure(e, attemptId, service);
    }
  }

  Future<void> _runStartConnectionHandshake(
    int attemptId,
    obd2lib.Obd2Service service,
  ) async {
    await obd2lib.Commands.ensureInitialized();
    if (!_isActiveConnectionAttempt(attemptId, service)) return;
    final info = await service.startConnection(
      timeout: 30.0,
      querySupportedPIDs: _querySupportedPids,
    );
    if (!_isActiveConnectionAttempt(attemptId, service)) return;
    _captureSupportedMode1Pids(info);
  }

  void _captureSupportedMode1Pids(obd2lib.ObdInfo info) {
    if (!_querySupportedPids) return;

    final pids = info.supportedPIDs;
    if (pids == null) return;

    _supportedMode1Pids = pids
        .map((cmd) => cmd.properties.command)
        .where((cmd) => cmd.startsWith('01'))
        .toSet();
    _hasSupportedMode1Snapshot = true;
    obdInfo('OBD-II connected successfully.', category: LogCategory.service);
  }

  void _handleStartConnectionFailure(
    Object e,
    int attemptId,
    obd2lib.Obd2Service service,
  ) {
    if (!_isActiveConnectionAttempt(attemptId, service)) {
      obdDebug(
        'Ignoring stale connection failure: $e',
        category: LogCategory.service,
      );
      return;
    }
    _setConnectionState(OBDConnectionState.failed);
    obdError('connect failed: $e', category: LogCategory.service);
  }

  bool _isActiveConnectionAttempt(int attemptId, obd2lib.Obd2Service service) =>
      attemptId == _connectionAttemptId && identical(service, _obdService);

  Future<bool> _ensureBluetoothPermissions() async {
    if (!Platform.isAndroid) return true;

    try {
      final statuses = await [
        Permission.bluetoothScan,
        Permission.bluetoothConnect,
      ].request();

      final missingBluetoothPerm = statuses.entries.any(
        (entry) => !entry.value.isGranted,
      );
      if (missingBluetoothPerm) {
        obdError(
          'bluetooth permission denied: '
          '${statuses.map((k, v) => MapEntry(k.toString(), v.toString()))}',
          category: LogCategory.service,
        );
        return false;
      }

      // Some Android devices still gate scan results behind location access.
      final locationStatus = await Permission.locationWhenInUse.status;
      if (!locationStatus.isGranted) {
        final requested = await Permission.locationWhenInUse.request();
        if (!requested.isGranted) {
          obdWarning('location permission denied: $requested',
              category: LogCategory.service);
        }
      }
    } catch (e) {
      obdError('permission request failed: $e', category: LogCategory.service);
      return false;
    }

    return true;
  }

  @override
  void disconnect() {
    _connectionAttemptId++;
    _obdService?.stopConnection();
    _dataSub?.cancel();
    _clearForTerminalState();
    _setConnectionState(OBDConnectionState.disconnected);
  }

  @override
  void updateConnectionDetails() {
    if (connectionState != OBDConnectionState.disconnected) disconnect();
    _rebuildService();
  }

  // ── State machine ────────────────────────────────

  void _handleServiceState(obd2lib.ConnectionState serviceState) {
    switch (serviceState) {
      case obd2lib.ConnectionState.disconnected:
        if (connectionState == OBDConnectionState.disconnected) {
          return;
        }
        obdDebug('Mirrored service disconnect to manager state and cleared data.',
            category: LogCategory.service);
        _clearForTerminalState();
        _setConnectionState(OBDConnectionState.disconnected);
        break;
      case obd2lib.ConnectionState.error:
        _clearForTerminalState();
        _setConnectionState(OBDConnectionState.failed);
        break;
      case obd2lib.ConnectionState.connecting:
        _setConnectionState(OBDConnectionState.connecting);
        break;
      case obd2lib.ConnectionState.connectedToAdapter:
        _setConnectionState(OBDConnectionState.connectedToAdapter);
        break;
      case obd2lib.ConnectionState.connectedToVehicle:
        if (connectionState != OBDConnectionState.connected) {
          obdInfo('Connected to vehicle.', category: LogCategory.service);
          _setConnectionState(OBDConnectionState.connected);
          _startContinuousUpdates(PidInterestRegistry.instance.interested);
        }
        break;
    }
  }

  /// Externally trigger the vehicle setup state (usually via log interception).
  void setSettingUpVehicle() {
    if (connectionState == OBDConnectionState.connectedToAdapter) {
      _setConnectionState(OBDConnectionState.settingUpVehicle);
    }
  }

  void _clearForTerminalState() {
    _dataSub?.cancel();
    _lastStreamingPids = {};
    _supportedMode1Pids = {};
    _hasSupportedMode1Snapshot = false;
    pidStats = {};
    fuelStatus = null;
    milStatus = null;
    troubleCodes = null;
    connectedPeripheralName = null;
    _pidStatsNotifier.value = {};
  }

  void _setConnectionState(OBDConnectionState s) {
    connectionState = s;
    _connStateNotifier.value = s;
    notifyListeners();
  }

  // ── Data streaming ───────────────────────────────

  void _startContinuousUpdates(Set<String> interested) {
    if (_obdService == null) return;
    final enabledNow = _filterInterestedPids(interested);
    if (enabledNow.isEmpty) return;
    if (setEquals(enabledNow, _lastStreamingPids)) return;

    _dataSub?.cancel();
    _lastStreamingPids = enabledNow;

    final isMetric = ConfigData.instance.units == MeasurementUnit.metric;
    final unit = isMetric
        ? obd2lib.MeasurementUnit.metric
        : obd2lib.MeasurementUnit.imperial;

    final commands = enabledNow
        .map((cmd) => obd2lib.Commands.allCommands[cmd])
        .whereType<obd2lib.ObdCommand>()
        .toList();

    if (commands.isEmpty) {
      obdInfo('No interested PIDs to monitor.', category: LogCategory.service);
      return;
    }

    obdInfo(
      'Starting continuous updates for ${commands.length} PIDs.',
      category: LogCategory.service,
    );

    _dataSub = _obdService!
        .startContinuousUpdates(pids: commands, unit: unit)
        .listen(_handleUpdateBatch);
  }

  void _restartContinuousUpdates(Set<String> interested) {
    _dataSub?.cancel();
    _lastStreamingPids = {};
    _startContinuousUpdates(interested);
  }

  Set<String> _filterInterestedPids(Set<String> interested) {
    if (!_querySupportedPids || !_hasSupportedMode1Snapshot) return interested;

    return interested.where((cmd) {
      if (!cmd.startsWith('01')) return true;
      return _supportedMode1Pids.contains(cmd);
    }).toSet();
  }

  void _handleUpdateBatch(
      Map<obd2lib.ObdCommand, obd2lib.DecodeResult> batch) {
    bool changed = false;

    for (final entry in batch.entries) {
      if (_processDecodeResult(entry.key.properties.command, entry.value)) {
        changed = true;
      }
    }

    if (changed) {
      _pidStatsNotifier.value = Map.from(pidStats);
      notifyListeners();
    }
  }

  bool _processDecodeResult(String cmdStr, obd2lib.DecodeResult decode) {
    if (cmdStr == '0103' && decode.codeResult != null) {
      fuelStatus = decode.codeResult;
      return true;
    }
    if (cmdStr == '0101' && decode.statusResult != null) {
      milStatus = decode.statusResult;
      return true;
    }
    if (cmdStr == '03' && decode.troubleCodes != null) {
      troubleCodes = decode.troubleCodes;
      return true;
    }
    if (decode.measurementResult != null) {
      final existing = pidStats[cmdStr];
      pidStats[cmdStr] = existing != null
          ? existing.copyWith(decode.measurementResult!)
          : PIDStats(pid: cmdStr, latest: decode.measurementResult!);
      return true;
    }
    return false;
  }


  // ── Stats helpers ─────────────────────────────────

  @override
  PIDStats? statsFor(String pidCommand) => pidStats[pidCommand];

  void _resetAllStats() {
    pidStats = {
      for (final e in pidStats.entries)
        e.key: PIDStats(pid: e.key, latest: e.value.latest)
    };
    _pidStatsNotifier.value = Map.from(pidStats);
  }

  // ── Stream helpers ────────────────────────────────

  Stream<T> _valueNotifierStream<T>(ValueNotifier<T> notifier) =>
      Stream.multi((c) {
        c.add(notifier.value);
        void l() => c.add(notifier.value);
        notifier.addListener(l);
        c.onCancel = () => notifier.removeListener(l);
      });

  Stream<T> _changeNotifierStream<T>(T Function() getter) =>
      Stream.multi((c) {
        c.add(getter());
        void l() => c.add(getter());
        addListener(l);
        c.onCancel = () => removeListener(l);
      });

  // ── Conversion helper ─────────────────────────────

  static obd2lib.ConnectionType _toLibConnType(ConnectionType t) {
    switch (t) {
      case ConnectionType.wifi:
        return obd2lib.ConnectionType.wifi;
      case ConnectionType.demo:
        return obd2lib.ConnectionType.demo;
      case ConnectionType.bluetooth:
        return obd2lib.ConnectionType.bluetooth;
    }
  }
}
